### ISSUE TITLE
Race condition in RedisCache leading to non-expiring keys

### ISSUE THREAD

**Issue by CaringDev** [dotnet/aspnetcore#59211]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

A race condition in https://github.com/dotnet/aspnetcore/blob/81a2bab8704d87d324039b42eb1bab0d977f25b8/src/Caching/StackExchangeRedis/src/RedisCache.cs#L233 leads to non-expiring keys, potentially Redis storage.

### Expected Behavior

Keys should be set to expire in Redis if configured accordingly.

### Steps To Reproduce

Change
```
await Task.WhenAll(
    cache.HashSetAsync(prefixedKey, fields),
    cache.KeyExpireAsync(prefixedKey, TimeSpan.FromSeconds(ttl.GetValueOrDefault()))
    ).ConfigureAwait(false);
```
to
```
await cache.KeyExpireAsync(prefixedKey, TimeSpan.FromSeconds(ttl.GetValueOrDefault()));
await cache.HashSetAsync(prefixedKey, fields);
```
to immediately see, that this will not set the expiration on the key.

### .NET Version

9.x

### Anything else?

Microsoft.Extensions.Caching.StackExchangeRedis @ 9.0.0 (8.0.11 is OK)

**Comment by mgravell** [comment]
I'm not using Bitnami yet (I'll need to investigate), but: initial attempts to reproduce a failure here at either the raw SE.Redis layer or the M.E.C.SER layer have been unsuccessful; so that we investigate and fix (as necessary) the correct problem, can you perhaps eyeball my test below, and/or see whether something similar to this can reproduce the problem in your setup?

At the API layer: the code is not wrong as far as I can see, in that SE.Redis is *meant* to allow this form of pipelining, and does indeed seem to be. Right now, I'm wondering whether, for example, it is the case that *Bitnami itself* doesn't respect pipelining.

```
using Microsoft.Extensions.Caching.Distributed;
using Microsoft.Extensions.DependencyInjection;
using StackExchange.Redis;
using System.Text;

var services = new ServiceCollection();
using var muxer = await ConnectionMultiplexer.ConnectAsync("127.0.0.1:6379");
var db = muxer.GetDatabase();
services.AddStackExchangeRedisCache(options =>
{
    options.ConnectionMultiplexerFactory = () => Task.FromResult<IConnectionMultiplexer>(muxer);
});
var options = new DistributedCacheEntryOptions { AbsoluteExpirationRelativeToNow = TimeSpan.FromMinutes(2) };
using var provider = services.BuildServiceProvider();
var cache = provider.GetRequiredService<IDistributedCache>();
var testkey = Guid.NewGuid().ToString();
cache.SetString(testkey, DateTime.UtcNow.ToString(), options);
var rand = new Random();
// prove setup correctly
Console.WriteLine($"cache: {cache.GetType().Name} ({cache.GetString(testkey)})");


int WORKER_COUNT = 20, PASSES = 1000, OPS_PER_PASS = 200;
var workers = Enumerable.Range(0, WORKER_COUNT).Select(StartWorker).ToArray();

var start = DateTime.UtcNow;
await Task.WhenAll(workers);
var end = DateTime.UtcNow;
Console.WriteLine($"Total ops: {WORKER_COUNT * PASSES * OPS_PER_PASS * 2}, time: {end - start}");
async Task StartWorker(int id)
{
    var skeys = new string[OPS_PER_PASS];
    for (int i = 0; i < skeys.Length; i++)
    {
        skeys[i] = Guid.NewGuid().ToString();
    }
    byte[] value = new byte[128];
    rand.NextBytes(value);
    var rkeys = Array.ConvertAll(skeys, key => (RedisKey)key);
    var ttls = new Task<TimeSpan?>[skeys.Length];

    async Task<int> CheckMissingExpirationsAsync()
    {
        for (int i = 0; i < ttls.Length; i++)
        {
            ttls[i] = db.KeyTimeToLiveAsync(rkeys[i]);
        }
        await Task.WhenAll(ttls);
        int failures = 0;
        for (int i = 0; i < ttls.Length; i++)
        {
            if ((await ttls[i]) is null) failures++;
        }
        return failures;
    }

    var counts = new Task<long>[OPS_PER_PASS];
    int totalFailures = 0;
    for (int pass = 0; pass < PASSES; pass++)
    {
        // reset and set via async
        await db.KeyDeleteAsync(rkeys);
        for (int i = 0; i < skeys.Length; i++)
        {
            await cache.SetAsync(skeys[i], value, options);
        }
        var failures = await CheckMissingExpirationsAsync();
        if (failures != 0)
        {
            Console.WriteLine($"Worker {id}, async, failures: {failures}");
            totalFailures += failures;
        }

        // reset and set via sync
        await db.KeyDeleteAsync(rkeys);
        for (int i = 0; i < skeys.Length; i++)
        {
            cache.Set(skeys[i], value, options);
        }
        failures = await CheckMissingExpirationsAsync();
        if (failures != 0)
        {
            Console.WriteLine($"Worker {id}, sync, failures: {failures}");
            totalFailures += failures;
        }
    }
    await db.KeyDeleteAsync(rkeys);
    Console.WriteLine($"Worker {id} complete, {PASSES} passes of {OPS_PER_PASS} async and {OPS_PER_PASS} sync, failures: {totalFailures}");
}
```

**Comment by mgravell** [comment]
I have zero knowledge of cloudfoundry; if there is anything you can do to help guide me producing this in that environment? my concern is that if I hack at it like so-many-monkeys, I end up with either a: something that doesn't work at all, or b: something that doesn't represent the failing scenario.

**Comment by CaringDev** [comment]
@mgravell thank you for the continued effort. I tried to run [your test suite](https://github.com/dotnet/aspnetcore/issues/59211#issuecomment-2511245228) with M.E.C.SER @ 9.0.0, SE.R @ 2.8.22 on the dev env and got (first line logged / redacted by me):

```
Connection string: redis-1:443,redis-2:443,redis-3:443,allowAdmin=True,password=<REDACTED>,ssl=True,abortConnect=False

cache: RedisCacheImpl (12/05/2024 17:07:26)
Unhandled exception. StackExchange.Redis.RedisCommandException: Multi-key operations must involve a single slot; keys can use 'hash tags' to help this, i.e. '{/users/12345}/account' and '{/users/12345}/contacts' will always be in the same slot
   at StackExchange.Redis.ConnectionMultiplexer.PrepareToPushMessageToBridge[T](Message message, ResultProcessor`1 processor, IResultBox`1 resultBox, ServerEndPoint& server) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 1969
   at StackExchange.Redis.ConnectionMultiplexer.TryPushMessageToBridgeAsync[T](Message message, ResultProcessor`1 processor, IResultBox`1 resultBox, ServerEndPoint& server) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 2007
   at StackExchange.Redis.ConnectionMultiplexer.ExecuteAsyncImpl[T](Message message, ResultProcessor`1 processor, Object state, ServerEndPoint server) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 2188
   at StackExchange.Redis.RedisBase.ExecuteAsync[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) in /_/src/StackExchange.Redis/RedisBase.cs:line 54
   at StackExchange.Redis.RedisDatabase.KeyDeleteAsync(RedisKey[] keys, CommandFlags flags) in /_/src/StackExchange.Redis/RedisDatabase.cs:line 862
   at Redis.Test.<>c__DisplayClass0_0.<<It>g__StartWorker|1>d.MoveNext() in /home/dev/src/Redis/Test.cs:line 74
--- End of stack trace from previous location ---
   at Redis.Test.It(String connectionString) in /home/dev/src/Redis/Test.cs:line 35
   at Program.<Main>$(String[] args) in /home/dev/src/Redis/Program.cs:line 30
   at Program.<Main>(String[] args)

Process finished with exit code 134.
```

**Comment by CaringDev** [comment]
> I have zero knowledge of cloudfoundry; if there is anything you can do to help guide me producing this in that environment? my concern is that if I hack at it like so-many-monkeys, I end up with either a: something that doesn't work at all, or b: something that doesn't represent the failing scenario.

(Un)fortunately I'm only "consumer" of that "cloud"... so I can't really help with that except to run random pieces of code on that environment. FYI: we're bound to migrate to RHOS (Red Hat OpenShift) next year (probably Q2) so we could also leave as is... as staying on 8.0.x for M.E.C.SER is not a dealbreaker for us (for now).

**Comment by mgravell** [comment]
ah, I didn't think "cluster", but: this is just the delete step; actually, you should be able to just delete the "delete" code, or unroll it - it isn't the important part of the test! perhaps replace with:

``` c#
foreach (var key in rkeys)
{
    db.KeyDelete(key, flags: CommandFlags.FireAndForget);
}
```

which will issue N individual delete operations, without waiting for them.

**Comment by CaringDev** [comment]
After replacing all three `await db.DeleteKeyAsync(rkeys)` with above loop and reducing workers / passes / ops I now get:

```
cache: RedisCacheImpl (12/16/2024 09:08:20)
Worker 2, async, failures: 4
Worker 1, async, failures: 2
Worker 0, async, failures: 6
Worker 2, sync, failures: 1
Worker 0, sync, failures: 1
Worker 1, sync, failures: 1
Worker 2, sync, failures: 2
Worker 1, sync, failures: 2
Worker 0, sync, failures: 3
Worker 2, async, failures: 1
Worker 1, async, failures: 1
Worker 0, async, failures: 2
Worker 0, sync, failures: 1
Worker 2 complete, 10 passes of 20 async and 20 sync, failures: 8
Worker 1 complete, 10 passes of 20 async and 20 sync, failures: 6
Worker 0 complete, 10 passes of 20 async and 20 sync, failures: 13
Total ops: 1200, time: 00:00:25.0325392
```

This is from my development VM (Ubuntu) against said Bitnami Redis cluster.

**Comment by mgravell** [comment]
Additional: now that we have the startings of a repro, there are some additional analysis tools we can use; specifically [RESP logging](https://stackexchange.github.io/StackExchange.Redis/RespLogging); this logs the actual data sent between client and server (bidirectionally). If we can get a repro *with RESP logging enabled*, that means we have a file snapshot that will give us a definitely answer between "1" and "2". If you are happy and able to run this and share the files with me (you may wish to nuke any `AUTH` data at the very top of the file, in the first few commands), I can parse it to see whether any commands were issued in an unexpected order. Because this happens at the transport layer rather than as part of the SE.Redis logic, it should be a true and accurate representation of what *actually* happened rather than what we *expected* to happen.

**Comment by CaringDev** [comment]
Thanks, I'll try to get my hands on a repro for the infrastructure... meanwhile, I captured some logs, replaced the actual PW with `pwd`, some.real.hostname with `host` and added .txt extensions due to GH filename restrictions.

```
cache: RedisCacheImpl (12/16/2024 11:03:01)
Worker 0, async, failures: 1
Worker 0, sync, failures: 1
Worker 0 complete, 1 passes of 10 async and 10 sync, failures: 2
Total ops: 20, time: 00:00:00.9568879
```
[redis-f35834bb-6845-0.redis.tcp-sni-svc.host^%443 Interactive 0.i.txt](https://github.com/user-attachments/files/18149318/redis-f35834bb-6845-0.redis.tcp-sni-svc.host.443.Interactive.0.i.txt)
[redis-f35834bb-6845-1.redis.tcp-sni-svc.host^%443 Interactive 2.i.txt](https://github.com/user-attachments/files/18149320/redis-f35834bb-6845-1.redis.tcp-sni-svc.host.443.Interactive.2.i.txt)
[redis-f35834bb-6845-2.redis.tcp-sni-svc.host^%443 Interactive 1.i.txt](https://github.com/user-attachments/files/18149322/redis-f35834bb-6845-2.redis.tcp-sni-svc.host.443.Interactive.1.i.txt)
[redis-f35834bb-6845-0.redis.tcp-sni-svc.host^%443 Interactive 0.ou.txt](https://github.com/user-attachments/files/18149319/redis-f35834bb-6845-0.redis.tcp-sni-svc.host.443.Interactive.0.ou.txt)
[redis-f35834bb-6845-0.redis.tcp-sni-svc.host^%443 Subscription 4.i.txt](https://github.com/user-attachments/files/18149325/redis-f35834bb-6845-0.redis.tcp-sni-svc.host.443.Subscription.4.i.txt)
[redis-f35834bb-6845-0.redis.tcp-sni-svc.host^%443 Subscription 4.ou.txt](https://github.com/user-attachments/files/18149323/redis-f35834bb-6845-0.redis.tcp-sni-svc.host.443.Subscription.4.ou.txt)
[redis-f35834bb-6845-1.redis.tcp-sni-svc.host^%443 Interactive 2.ou.txt](https://github.com/user-attachments/files/18149326/redis-f35834bb-6845-1.redis.tcp-sni-svc.host.443.Interactive.2.ou.txt)
[redis-f35834bb-6845-1.redis.tcp-sni-svc.host^%443 Subscription 5.i.txt](https://github.com/user-attachments/files/18149321/redis-f35834bb-6845-1.redis.tcp-sni-svc.host.443.Subscription.5.i.txt)
[redis-f35834bb-6845-1.redis.tcp-sni-svc.host^%443 Subscription 5.ou.txt](https://github.com/user-attachments/files/18149317/redis-f35834bb-6845-1.redis.tcp-sni-svc.host.443.Subscription.5.ou.txt)
[redis-f35834bb-6845-2.redis.tcp-sni-svc.host^%443 Interactive 1.ou.txt](https://github.com/user-attachments/files/18149327/redis-f35834bb-6845-2.redis.tcp-sni-svc.host.443.Interactive.1.ou.txt)
[redis-f35834bb-6845-2.redis.tcp-sni-svc.host^%443 Subscription 3.i.txt](https://github.com/user-attachments/files/18149328/redis-f35834bb-6845-2.redis.tcp-sni-svc.host.443.Subscription.3.i.txt)
[redis-f35834bb-6845-2.redis.tcp-sni-svc.host^%443 Subscription 3.ou.txt](https://github.com/user-attachments/files/18149324/redis-f35834bb-6845-2.redis.tcp-sni-svc.host.443.Subscription.3.ou.txt)

**Comment by mgravell** [comment]
We're not looking for invalid fragments; I see the same, but I think the problem *there* is that the files have been manually edited by a text editor, and that text editor has corrupted the data. The main reason I guess this is from:

```
*2
$4
AUTH
$32
pwd
....
```

which is ... just wrong. I'm guessing that the text editor has borked all the non-printable characters, making a mess of things. However, in the data that I can eyeball, the order is always correctly `HMSET` then `EXPIRE`, which again strongly suggests that this is *not* a client library problem. My main hunch here continues to be an exotic proxy server that is spreading load between multiple proxy-to-redis connections, without realising that this fundamentally breaks pipelining - an expected Redis use-case.

**Comment by mgravell** [comment]
To be clear: what I did here was

- manually search in the file fie each `EXPIRE` entry
- select the guid for that `EXPIRE`, which in my editor automatically shows other occurrences
- look (just eyeballs) for where else it occurs, and in each case: it is in a `HMSET` shortly before the `EXPIRE`

**Comment by CaringDev** [comment]
Meanwhile I can confirm that `LoggingTunnel.ValidateAsync` does _not_ find any issues, even for (unaltered) traces that exhibit the problem at hand.

**Comment by mgravell** [comment]
We haven't written the code to *detect* the specific problem yet ;p `ValidateAsync` just checks for RESP validity - it doesn't do any semantic analysis; I can't run this locally because of the file corruption, however something like this might work for you, or it might need tweaking - I can't tell without being able to run it; what it is doing is:

- for each file, replay the conversation
- find the outbound `HMSET` and `EXPIRE` messages
- for `HMSET`, record that we have a key by adding it to `keys`
- for `EXPIRE`, mark that key as handled by removing it from `keys`
- check the balance!

Code:

``` c#
foreach (var outFile in Directory.GetFiles(@"C:\Code\RedisLog", "*Interactive*.out"))
{
    var inFile = Path.ChangeExtension(outFile, "in");
    if (File.Exists(outFile))
    {
        Console.WriteLine(Path.GetFileNameWithoutExtension(outFile));
        using var outStream = File.OpenRead(outFile);
        using var inStream = File.OpenRead(inFile);
        int validated = 0, tooSoon = 0;

        HashSet<string> keys = new HashSet<string>();
        try
        {
            await LoggingTunnel.ReplayAsync(outStream, inStream, (req, resp) =>
            {
                Console.WriteLine(req.Length);
                if (req.Length > 1)
                {
                    string cmd = ((string)req[0]!);
                    switch(cmd.ToUpperInvariant())
                    {
                        case "HMSET":
                            keys.Add((string)req[1]!);
                            break;
                        case "EXPIRE":
                            if (keys.Remove((string)req[1]!))
                            {
                                validated++;
                            }
                            else
                            {
                                tooSoon++;
                                Console.WriteLine("Key either never existed or expired twice");
                            }
                            break;
                    }
                    Console.WriteLine();
                }
            });
            Console.WriteLine($"{validated} valid, {keys.Count} left unexpired, {tooSoon} expired too soon");
        }
        catch (Exception ex)
        {
            Console.WriteLine(ex.Message);
        }
    }
}
```

**Comment by mgravell** [comment]
The other thing we could do is that code is: in `EXPIRE`, check `(bool)resp` - it should be `true`:

``` c#
int serverSaidNo = 0;
//...
                        case "EXPIRE":
                            if (!(bool)resp) serverSaidNo++;
                            if (keys.Remove((string)req[1]!))
                            {
                                validated++;
                            }
                            else
                            {
                                tooSoon++;
                            }
                            break;
```

and output `serverSaidNo` at the end; we *hope* it is `0`, but it'll be interesting to see if it is non-zero

**Comment by CaringDev** [comment]
Thought about the counting already but prioritized dinner with my wife :-)
Now it gets interesting (enabled RESP3, disabled CLIENT SETINFO [due to Redis < 7.2], slightly adapted output and added MOVE detection):

```
redis-7c2b3133-0bf4-0.redis.tcp-sni-svc.host^%443 Interactive 0
7, 2, 3, 2, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 8, 3, 8, 3, 8, 3, 2, 2, 2, 2, 2, 2, 8, 3, 8, 3, 8, 3, 2, 2, 2, 2, 2, 1
6 valid, 0 left unexpired, 0 expired too soon, 0 rejected by server, 0 moved
redis-7c2b3133-0bf4-1.redis.tcp-sni-svc.host^%443 Interactive 2
7, 2, 3, 2, 3, 3, 2, 2, 2, 2, 2, 2, 2, 8, 3, 5, 2, 8, 2, 2, 2, 2, 3, 8, 3, 8, 3, 8, 3, 8, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 8, 3, 8, 3, 8, 3, 8, 3, 8, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1
11 valid, 0 left unexpired, 0 expired too soon, 1 rejected by server, 0 moved
redis-7c2b3133-0bf4-2.redis.tcp-sni-svc.host^%443 Interactive 1
7, 2, 3, 2, 3, 3, 2, 2, 2, 2, 2, 2, 2, 3, 2, 3, 2, 2, 3, 8, 3, 8, 3, 2, 2, 2, 2, 2, 8, 3, 8, 3, 2, 2, 2, 2, 1
4 valid, 0 left unexpired, 3 expired too soon, 0 rejected by server, 3 moved
```

Shall I send the files by e.g. e-mail?

**Comment by mgravell** [comment]
Ah, the `MOVED` is interesting. I wonder if that's a big missing piece here - as in: maybe the moved is part of how things are getting out-of-order; that's certainly something I can investigate, although normally that's basically impossible to repro on demand - `MOVED` is usually very rare! In the `MOVED` case, I believe the "too soon" is probably related to detection criteria.

It also occurred: in the "weird proxies" case: I wonder if we can use the return of the expire, i.e.

- pipeline both
- await both, check the result
- if `false`, reissue both, but awaiting in turn

This means we only pay latency if there's an actual problem..certainly worth a try.

**Comment by mgravell** [comment]
On the topic of files by email; you're welcome to - probably marcgravell@microsoft.com is best. To be clear: I have no malign interest in your data or servers, but if you send me files contain things like valid auth tokens, you should probably rotate them. If you have a hex editor handy, you can always overwrite those bytes at the binary level (stopping before the \r\n that is the terminator); however, text editors are probably going to keep causing corruption problems!

**Comment by CaringDev** [comment]
Thanks... to get more meaningful data I turned up the volume again: 3 workers, 10 passes, 100 ops:

```
Worker 0 complete, 10 passes of 100 async and 100 sync, failures: 247
Worker 1 complete, 10 passes of 100 async and 100 sync, failures: 278
Worker 2 complete, 10 passes of 100 async and 100 sync, failures: 276
```

and the analysis shows

```
redis-7c2b3133-0bf4-2.redis.tcp-sni-svc.host:443 Interactive 1
1952 valid, 110 left unexpired, 476 expired too soon, 250 rejected by server, 408 moved
redis-7c2b3133-0bf4-0.redis.tcp-sni-svc.host:443 Interactive 0
2839 valid, 63 left unexpired, 532 expired too soon, 363 rejected by server, 1171 moved
redis-7c2b3133-0bf4-1.redis.tcp-sni-svc.host:443 Interactive 2
1813 valid, 109 left unexpired, 362 expired too soon, 185 rejected by server, 394 moved
```

The logs (with altered PWD but telling hostnames :-)) are on their way...

**Comment by mgravell** [comment]
OK; this is getting super interesting! We have more hints, and I *suspect* there may be an immediate workaround / mitigation - read more below!

---

So: what is happening here seems heavily related to the `-MOVED`; as I mentioned previously, `-MOVED` ***should*** be *incredibly* rare, since SE.Redis is a cluster-aware client and performs routing in the library - the only times we *expect* to see this is during a re-shard operation.

There is no re-shard operation in progress in this scenario, and I can confirm that SE.Redis has routed commands incorrectly; in the trace, `CLUSTER NODES` is reporting, for example:

``` txt
8992b30e1499e2cf0bc86609d752c4f3f49c2199 AAAAA master - 0 1734422383728 2 connected 5461-10922
c9ef1500c2acc37273e5b04275f7e5620b796fff BBBBB myself,master - 0 1734422382000 1 connected 0-5460
753343875ffe163a64284b36d7743861f2ef1506 CCCCC master - 0 1734422384732 3 connected 10923-16383
```

the `myself` means we're talking to the `BBBBB` node (names removed for privacy), which is serving slots 0-5460; on the same connection, we then see, for example, commands issued for key `c96dd826-29dd-42fa-8721-91dd36b08626`; running the SE.Redis version of the slot algorithm, this generates 10336, which should be handled by node `AAAAA`; and sure enough, the response is:

``` txt
-MOVED 10336 AAAAA
```

which means that the server agrees with our slot calculation - we just sent it to the wrong node! This is ... weird and concerning.

In theory, SE.Redis should still follow these redirects, reissuing the commands to the indicated server. I wonder, then, if the *immediate* cause of a problem here is some timing issue when reissuing `-MOVED` commands, and *that* is where they are getting out of order. I need to investigate and resolve this.

Separately, there is the question of why we routed it incorrectly in the first place! One possible thought here is that the problem is coming from RESP3, and in particular the fact that in RESP3, Redis responds to `CLUSTER NODES` with a *verbatim string*; this is *like* a bulk string from RESP2, but it has a prefix on the data; the above quote was a white lie to lay a trail of suspense and misdirection, but the *actual* response to `CLUSTER NODES` was:

``` txt
txt:8992b30e1499e2cf0bc86609d752c4f3f49c2199 AAAAA master - 0 1734422383728 2 connected 5461-10922
c9ef1500c2acc37273e5b04275f7e5620b796fff BBBBB myself,master - 0 1734422382000 1 connected 0-5460
753343875ffe163a64284b36d7743861f2ef1506 CCCCC master - 0 1734422384732 3 connected 10923-16383
```

So: I wonder if the *secondary reason* behind the problem is that our RESP3 handling for `CLUSTER NODES` is not handling this prefix correctly. I will investigate this and resolve ASAP, but it would help if you could confirm this by *removing* the RESP3 switch for now (RESP3 is opt-in, so you must be doing something to actively enable it, probably on the configuration string). I would greatly appreciate it if you could let me know whether dropping back to RESP2 mitigates this problem - if it does, that most likely confirms that our `CLUSTER NODES` handling in RESP3 is broken.

To summarise next steps:

- if possible, @CaringDev to see whether RESP2 has the same problem (remove RESP3 enable step)
- @mgravell separately to investigate `CLUSTER NODES` handling in RESP3 (linked to ^^^)
- @mgravell to look at whether there's an out-of-order risk with `-MOVED` handling

Copying @NickCraver @philon-msft and @adityamandaleeka for visibility.

**Comment by mgravell** [comment]
Initial tests suggest that the RESP3 handling of `CLUSTER NODES` *is not* the cause here; the code correctly removes the prefix, so; it is possibly (likely, even) that switching to RESP2 will have no effect, and that the problem is something deeper in the topology code.

**Comment by mgravell** [comment]
Current hypothesis: for some reason, *in your topology* we are considering the redis endpoints as unselectable; in that case, our "we have nothing to lose" fallback is basically "send it anywhere, and see what happens"; this would also explain why there is a timing problem with `-MOVED` - if we're sending two commands *anywhere*, and relying on `-MOVED`, then the order is determined by how those two individual messages get processed *potentially by two different servers*, possibly with `-MOVED` redirects involving a second message, or possibly processed immediately.

So: I think we can *defence-in-depth* mitigate against this in the caching client by (as suggested previously) checking the result of the second operation and re-issuing with dual `await`, however: it would be great to solve the real problem here!

 I *wonder* whether the *real* problem is something to do with IP vs host-based routing; will need to look.

**Comment by CaringDev** [comment]
RESP2 shows the same symptoms.

Regarding IPs / hosts / host names: there's split DNS at play here... our public-looking host names are not publicly resolvable, but are internally. Specifically, indeed, CLUSTER NODES advertises non-routable IPs and the hostnames all point to the same IP (10.176....). What's more: `muxer.GetServers()` returns 6 entries, 3 of which are non-reachable IP endpoints:

```
// Console.WriteLine(server.EndPoint)
Unspecified/redis-host-0:443
Unspecified/redis-host-0:443
Unspecified/redis-host-0:443
10.42...X:443
10.42...Y:443
10.42...Z:443
```

**Comment by CaringDev** [comment]
FWIW: with M.E.C.SER 8.0.11 I get only a handful of `failures`.
(Replaying is pointless as the current logic is not suitable for the old mechanism... as is it does not detect any `MOVED`s though).

**Comment by mgravell** [comment]
I would be very surprised if there are no `-MOVED`; I suspect the problem is more: the routing table is borked because of the host vs ip problem, and *when using Lua*, the `-MOVED` redirection is masking that; basically: when we look the item up *by slot*, we try to send it to the IP, which turns out to be unavailable, so we send it to a random *available* node which is the host-based ones; we then get a `-MOVED` which *indirectly adds to the routing table*, and we route *there*, but that only impacts the *current* request - other items routed by key continue to try be sent via IP.

Basically: I think what we need to do here is to replicate *something like* this setup in SE.Redis using host-based routing in Redis 7 rather than IP-based routing, and see if we can make it work there. Or in other words: I do not believe that the relatively new feature of host-based routing in Redis 7 clusters has been implemented by the SE.Redis client, and: it is getting into a huge mess. Historically, it expects routable IP-based routing.

**Comment by mgravell** [comment]
On an internal chat, @philon-msft echo-d the sentiment that the addresses returned from `CLUSTER NODES` should be routable by the client. If they're not, that would 100% explain these symptoms because:

- initially, the set of nodes will consist of the host-based address in the connection string
- `CLUSTER NODES` will add some IP-based endpoints, but they will all fail to connect
- when routing commands, the library will fail to find a resolvable live endpoint, and instead use a "hail mary" strategy of sending it to the only live endpoint it has - where-ever we got to via the connection string
- this will generate `-MOVED` that *seems to include the host on redis 7*
- we redirect via the `-MOVED`, adding extra endpoints as needed
- which are then included in the "hail mary" routing, but will not be used for the primary routing
- chaos, bad things, human sacrifice, dogs and cats living together, MASS HYSTERIA!

So; long term, we can see whether SE.Redis should improve support for host-based `CLUSTER` routing, but *right now*, the short version would be: make the `CLUSTER NODES` advertised endpoints routable.

@CaringDev  thoughts?

**Comment by mgravell** [comment]
@CaringDev  if I can refer you to [`CLUSTER NODES`](https://redis.io/docs/latest/commands/cluster-nodes/), **emphasis mine**:

> 2. `ip:port@cport`: The node address ***that clients should contact to run queries***, along with the used cluster bus port. `:0@0` can be expected when the address is no longer known for this node ID, hence flagged with `noaddr`.
> 3. `hostname`: A human readable string that can be configured via the `cluster-annouce-hostname` setting. The max length of the string is 256 characters, excluding the null terminator. The name can contain ASCII alphanumeric characters, '-', and '.' only.

It seems fairly explicit that the IP endpoint is the correct case, and that the text name (which is configurable) is intended for reference only. This is similarly the case in [`CLUSTER SHARDS`](https://redis.io/docs/latest/commands/cluster-shards/), which has a different layout but again specifies that the IP and port is the expected use-case, and that the hostname is optional and may or may not be included.

**Comment by mgravell** [comment]
I have logged the client improvement logic ^^^

**Comment by BrennanConroy** [comment]
Seems like this isn't an issue with aspnetcore? Can we close?
